pkg/controller/podgc/gc_controller.go

```diff
@@ -316,26 +316,59 @@ func (gcc *PodGCController) markFailedAndDeletePod(ctx context.Context, pod *v1.
 	return gcc.markFailedAndDeletePodWithCondition(ctx, pod, nil)
 }
 
-func (gcc *PodGCController) markFailedAndDeletePodWithCondition(ctx context.Context, pod *v1.Pod, condition *v1.PodCondition) error {
+func (gcc *PodGCController) markFailedAndDeletePodWithCondition(ctx context.Context, pod *v1.Pod, condition *corev1apply.PodConditionApplyConfiguration) error {
 	klog.InfoS("PodGC is force deleting Pod", "pod", klog.KRef(pod.Namespace, pod.Name))
 	if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
-		newStatus := pod.Status.DeepCopy()
+
+		fieldManager := "PodGC"
+
+		// Extact the pod status as PodGC may or may not own the pod phase, if
+		// it owns the phase then we need to send the field back if the condition
+		// is added.
+		podApply, err := corev1apply.ExtractPodStatus(pod, fieldManager)
+		if err != nil {
+			return nil
+		}
+
+		// Set the status in case PodGC does not own any status fields yet
+		if podApply.Status == nil {
+			podApply.WithStatus(corev1apply.PodStatus())
+		}
+
 		updated := false
 		if condition != nil {
-			updated = apipod.UpdatePodCondition(newStatus, condition)
+			updatePodCondition(podApply.Status, condition)
+			updated = true
 		}
 		// Mark the pod as failed - this is especially important in case the pod
 		// is orphaned, in which case the pod would remain in the Running phase
 		// forever as there is no kubelet running to change the phase.
 		if pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
-			newStatus.Phase = v1.PodFailed
+			podApply.Status.WithPhase(v1.PodFailed)
 			updated = true
 		}
 		if updated {
-			if _, _, _, err := utilpod.PatchPodStatus(ctx, gcc.kubeClient, pod.Namespace, pod.Name, pod.UID, pod.Status, *newStatus); err != nil {
+			if _, err := gcc.kubeClient.CoreV1().Pods(pod.Namespace).ApplyStatus(ctx, podApply, metav1.ApplyOptions{FieldManager: "PodGC", Force: true}); err != nil {
 				return err
 			}
 		}
 	}
 	return gcc.kubeClient.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, *metav1.NewDeleteOptions(0))
 }
+
+func updatePodCondition(podStatusApply *corev1apply.PodStatusApplyConfiguration, condition *corev1apply.PodConditionApplyConfiguration) {
+	if conditionIndex, _ := findPodConditionApplyByType(podStatusApply.Conditions, *condition.Type); conditionIndex < 0 {
```

Comment:
@jpbetz is there an easier way to do this? If not maybe we should consider modifying the generated types?

Reply:
should we be looking at all the conditions (including the ones that are not owned by this controller)?

Reply:
In the current implementation I only look at the pod conditions owned by PodGC and I think this is enough.

If the condition is absent or owned by another controller, then we don't extract it, so PodGC will add it and it will be added or updated at the server side (thanks for Force: true). If the condition is owned by PodGC, then it will be extracted and we will update it in the list.

Reply:
> @jpbetz is there an easier way to do this? If not maybe we should consider modifying the generated types?

No, but we could offer something. The problem here is that the flow is using a extract/modify-in-place/apply and so it's not possible to just append some conditions and let the merge happen using the merge keys. We could offer a utility to lookup by key like `podStatusApply.WithCondition(typeKey string, condition)` that would set the condition with the key (using the same approach taken here, finding it and replacing it if it exists, otherwise appending it).

Reply:
> We could offer a utility to lookup by key and then set the condition

[pkg/api/meta.SetStatusCondition](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#SetStatusCondition) also has some logic for LastTransitionTime.

Reply:
Having the utility function seems like a good idea that allows to simplify the code in the future. As for the logic about updating LastTransitionTime (SetStatusCondition skips the update the if there is no status change). However, for SSA it might be tricky to replicate cause it would also depend on the ownership of the existing condition (as extraction depends on it).